File: typescript/src/booking/application/usecases/bookingPlacing.ts
// Caso d'uso per l'inserimento di una nuova prenotazione.

import type { UseCase } from '@/common/application/usecase.js';
import type { EventPublisher } from '@/common/application/events/eventPublisher.js';
import { Booking } from '../../domain/booking/booking.js';
import type { BookingEvent } from '../../domain/events/bookingEvent.js';
import { BookingPlaced, bookingPlaced } from '../../domain/events/bookingPlaced.js';
import type { BookingRepository } from '../../domain/ports/bookingRepository.js';
import { PlaceBooking } from '../commands/placeBooking.js';
import { requireArgument } from '@/common/utils/requireArgument.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class BookingPlacing implements UseCase<PlaceBooking, BookingPlaced> {
  private readonly _bookingRepository: BookingRepository;
  private readonly _eventPublisher: EventPublisher<BookingEvent>;

  constructor(bookingRepository: BookingRepository, eventPublisher: EventPublisher<BookingEvent>) {
    requireDependency(bookingRepository, "bookingRepository");
    requireDependency(eventPublisher, "eventPublisher");
    this._bookingRepository = bookingRepository;
    this._eventPublisher = eventPublisher;
  }

  invoke(command: PlaceBooking): BookingPlaced {
    requireArgument(command, 'command');

    const booking = Booking.place(command.aggregateId, command.description, command.giftCardId);

    this._bookingRepository.save(booking);

    const placed = bookingPlaced(booking.id(), command.amount, booking.description(), booking.giftCardId());
    this._eventPublisher.publish(placed);

    return placed;
  }
}

File: java/src/main/java/it/giannibombelli/wsc2026/booking/application/usecases/BookingPlacing.java
package it.giannibombelli.wsc2026.booking.application.usecases;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.booking.application.commands.PlaceBooking;
import it.giannibombelli.wsc2026.booking.domain.booking.Booking;
import it.giannibombelli.wsc2026.booking.domain.events.BookingEvent;
import it.giannibombelli.wsc2026.booking.domain.events.BookingPlaced;
import it.giannibombelli.wsc2026.booking.domain.ports.BookingRepository;
import it.giannibombelli.wsc2026.common.application.UseCase;
import it.giannibombelli.wsc2026.common.application.events.EventPublisher;

import static java.util.Objects.requireNonNull;

public final class BookingPlacing implements UseCase<PlaceBooking, BookingPlaced> {
    private final BookingRepository bookingRepository;
    private final EventPublisher<BookingEvent> eventPublisher;

    public BookingPlacing(BookingRepository bookingRepository,
                          EventPublisher<BookingEvent> eventPublisher) {
        this.bookingRepository = requireNonNull(bookingRepository);
        this.eventPublisher = requireNonNull(eventPublisher);
    }

    @Override
    public BookingPlaced invoke(PlaceBooking cmd) {
        Require.requireArgument(cmd, "command");

        Booking booking = Booking.place(cmd.aggregateId(), cmd.description(), cmd.giftCardReference());

        bookingRepository.save(booking);

        BookingPlaced bookingPlaced = new BookingPlaced(booking.id(), cmd.amount(), booking.description(), booking.giftCardReference().value());
        eventPublisher.publish(bookingPlaced);

        return bookingPlaced;
    }
}
